plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java

```diff
@@ -499,6 +509,36 @@ public MongoCursor<Document> execute(MongoTableHandle tableHandle, List<MongoCol
         return iterable.iterator();
     }
 
+    /**
+     * Creates a set of sufficient columns for the input projected columns. For example,
+     * if input {@param columns} include columns "a.b" and "a.b.c", then they will be projected from a single column "a.b".
+     */
+    public static List<MongoColumnHandle> getSufficientColumns(List<MongoColumnHandle> columnHandles)
+    {
+        List<MongoColumnHandle> sufficientColumns = new ArrayList<>();
+        for (MongoColumnHandle column : columnHandles) {
+            String qualifiedName = column.getQualifiedName();
+            List<MongoColumnHandle> columnsToRemove = new ArrayList<>();
+            boolean isAddable = true;
+
+            for (MongoColumnHandle existingColumn : sufficientColumns) {
+                String name = existingColumn.getQualifiedName();
+                if (name.startsWith(qualifiedName + ".")) {
+                    columnsToRemove.add(existingColumn);
+                }
+                else if (qualifiedName.startsWith(name + ".")) {
+                    isAddable = false;
+                    break;
+                }
+            }
+            sufficientColumns.removeAll(columnsToRemove);
+            if (isAddable) {
+                sufficientColumns.add(column);
+            }
+        }
+        return sufficientColumns;
+    }
+
```

Comment:
And do we need this simplification ? Like for a query like `SELECT row.field_1, row.field_1.nested_field`  MongoDB will simplifies it at their end right ? Since we share the `Document` across all the columns (incl. nested columns) it is kind of an overhead at our end. WDYT ?

Reply:
As discussed offline, `getSufficientColumns` is not needed. Even if `root` and `root.leaf1`  is the projected column then mongo client takes care to return the document with `root` only.

![image (3)](https://github.com/trinodb/trino/assets/11939923/c8c237a3-b66b-4f84-9b0a-d534290e2672)

![image (4)](https://github.com/trinodb/trino/assets/11939923/42e8fe9d-a26a-4f1c-bc52-326096e2c41a)

Reply:
I was partially incorrect in the previous comment.

if the order of the projection fields matters here.

-  output order -> (`root.leaf1`, `root`)
   - mongo client returns `root` document
-  output order -> (`root`, `root.leaf1`)
   - mongo client returns `root.leaf1` document

So It makes it necessary to provide projection in a manner where no filed isleft out.

<img width="826" alt="image" src="https://github.com/trinodb/trino/assets/11939923/593c3e08-67da-4764-ac13-5c5f70f773ec">

Reply:
> In that case - is it an issue in MongoDB client or ?

Yes, MongoDB somehow overrides the parent column if child column is later than parent in the projection call.

Reply:
>Like we can use an Ordering

@Praveen2112, I tried this approach.

When the table definition is

```
(id INT, row1_t ROW(f1 INT, f2 INT, row2_t ROW (f1 INT, f2 INT, row3_t ROW(f1 INT, f2 INT))))
```

and, projected column is
```
id, row1_t.row2_t.row3_t.f2, CAST(row1_t AS JSON)
```

then mongoclient returns the document without `row1_t.row2_t.f1`, `row1_t.row2_t.f2`, and `row1_t.row2_t.row3_t.f1`

<img width="1139" alt="image" src="https://github.com/trinodb/trino/assets/11939923/b404648d-8ecc-41b9-bbad-f32dfa736dd6">

Reply:
It seems that when we have ancestor and child as part of projection then, only one level of the ancestor document gets returns from mongoclient.

Reply:
Thanks @Praveen2112 for pointing out the doc offline. https://www.mongodb.com/docs/manual/reference/limits/#mongodb-limit-Projection-Restrictions


Here is the reason: `Path Collision: Embedded Documents and Its Fields`

`
If the projection of the embedded document comes before the projection any of its fields, MongoDB projects the specified field or fields. For example, the projection document { "size.uom": 1, size: 1, "size.h": 1 } produces the same result as the projection document { "size.uom": 1, "size.h": 1 }.
`

Reply:
`getSufficientColumns` is renamed to `projectSufficientColumns`. This method is required for path collision cases.